shell/platform/glfw/system_utils_test.cc

```diff
@@ -0,0 +1,321 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/glfw/system_utils.h"
+
+#include <cstdlib>
+
+#include "gtest/gtest.h"
+
+namespace flutter {
+namespace {
+
+// This is a helper for setting up the different environment variables to
+// specific strings, calling GetPreferredLanguageInfo, and then restoring those
+// environment variables to any previously existing values.
+std::vector<LanguageInfo> SetAndRestoreLanguageAroundGettingLanguageInfo(
+    const char* language,
+    const char* lc_all,
+    const char* lc_messages,
+    const char* lang) {
+  std::string old_language;
+  const char* language_env = getenv("LANGUAGE");
+  bool language_was_set = language_env != nullptr;
+  if (language_was_set) {
+    old_language = std::string(language_env);
+  }
+  std::string old_lc_all;
+  const char* lc_all_env = getenv("LC_ALL");
+  bool lc_all_was_set = lc_all_env != nullptr;
+  if (lc_all_was_set) {
+    old_lc_all = std::string(lc_all_env);
+  }
+  std::string old_lc_messages;
+  const char* lc_messages_env = getenv("LC_MESSAGES");
+  bool lc_messages_was_set = lc_messages_env != nullptr;
+  if (lc_messages_was_set) {
+    old_lc_messages = std::string(lc_messages_env);
+  }
+  std::string old_lang;
+  const char* lang_env = getenv("LANG");
+  bool lang_was_set = lang_env != nullptr;
+  if (lang_was_set) {
+    old_lang = std::string(lang_env);
+  }
+
+  if (language != nullptr) {
+    setenv("LANGUAGE", language, 1);
+  } else if (language_was_set) {
+    unsetenv("LANGUAGE");
+  }
+  if (lc_all != nullptr) {
+    setenv("LC_ALL", lc_all, 1);
+  } else if (lc_all_was_set) {
+    unsetenv("LC_ALL");
+  }
+  if (lc_messages != nullptr) {
+    setenv("LC_MESSAGES", lc_messages, 1);
+  } else if (lc_messages_was_set) {
+    unsetenv("LC_MESSAGES");
+  }
+  if (lang != nullptr) {
+    setenv("LANG", lang, 1);
+  } else if (lang_was_set) {
+    unsetenv("LANG");
+  }
+
+  std::vector<LanguageInfo> languages = GetPreferredLanguageInfo();
+
+  if (language_was_set) {
+    setenv("LANGUAGE", old_language.c_str(), 1);
+  }
+  if (lc_all_was_set) {
+    setenv("LC_ALL", old_lc_all.c_str(), 1);
+  }
+  if (lc_messages_was_set) {
+    setenv("LC_MESSAGES", old_lc_messages.c_str(), 1);
+  }
+  if (lang_was_set) {
+    setenv("LANG", old_lang.c_str(), 1);
+  }
+
```

Comment:
Optional nit: it seems like you could simplify this code substantially by using a map of values to set, and a map of previous values, so that you wouldn't need four copies of all of the logic and could instead just iterate through maps.

Reply:
Good idea, done.